log info on invalid --output-version

juanvallejo opened the pull request:
**Release note**:

``` release-note
release-note-none
```

Object versions default to the current version (v1) when a specified
`--output-version` is invalid. This patch logs a warning when this is
the case. Cases affected are all commands with the `--output-version`
option, and anywhere runtime objects are converted to versioned objects.

**Example**

```
$ kubectl get pod <mypod> -o json --output-version=invalid
W1013 17:24:16.810278   26719 result.go:238]  info: the output version
specified (invalid) is invalid, defaulting to v1
{
        "kind": "Pod",
            "apiVersion": "v1",
                "metadata": {
                            "name": "mypod",
                                    "namespace": "test",
...
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34763)

<!-- Reviewable:end -->

AdoHe commented:
@juanvallejo how you expect an empty version behavior?

juanvallejo commented:
@AdoHe 
> how you expect an empty version behavior?

Thanks for catching this, addressed here https://github.com/kubernetes/kubernetes/pull/34763/commits/00cfd1f980ae77b85d7d54418870e3c10a03c2e3

brendandburns commented on pkg/kubectl/resource/result.go:
```diff
@@ -228,6 +228,19 @@ func AsVersionedObject(infos []*Info, forceList bool, version unversioned.GroupV
+	// validSpecifiedVersion resolves to true if the version passed to this function matches the
+	// version assigned to the converted object
+	actualVersion := object.GetObjectKind().GroupVersionKind()
+	validSpecifiedVersion := (actualVersion.Version == version.Version)
```
I'd just inline this in the if statement.

brendandburns commented on pkg/kubectl/resource/result.go:
```diff
@@ -228,6 +228,19 @@ func AsVersionedObject(infos []*Info, forceList bool, version unversioned.GroupV
+		if len(actualVersion.Version) > 0 {
+			defaultVersionInfo = fmt.Sprintf(", defaulting to %q", actualVersion.Version)
+		}
+		fmt.Errorf(" info: the output version specified is invalid%s\n", defaultVersionInfo)
```
space between "invalid" and %s

juanvallejo commented:
@brendandburns Thanks for the feedback! Comments addressed https://github.com/kubernetes/kubernetes/pull/34763/commits/81fab4618b5ecc129e359f56e9456be9f1f1eb1a

juanvallejo commented on pkg/kubectl/resource/result.go:
```diff
@@ -228,6 +228,18 @@ func AsVersionedObject(infos []*Info, forceList bool, version unversioned.GroupV
+		if len(actualVersion.Version) > 0 {
+			defaultVersionInfo = fmt.Sprintf("Defaulting to %q", actualVersion.Version)
+		}
+		fmt.Errorf(" info: the output version specified is invalid. %s\n", defaultVersionInfo)
```
Thanks for catching this, I have updated it to use `glog.V(1)` instead

juanvallejo commented:
@k8s-merge-robot keep open

ncdc commented:
@juanvallejo squash please

ncdc commented:
ok to test

ncdc commented:
@juanvallejo could you please update bazel and test owners if necessary?

ncdc commented on pkg/kubectl/resource/result.go:
```diff
@@ -228,6 +229,18 @@ func AsVersionedObject(infos []*Info, forceList bool, version unversioned.GroupV
+		if len(actualVersion.Version) > 0 {
+			defaultVersionInfo = fmt.Sprintf("Defaulting to %q", actualVersion.Version)
+		}
+		glog.V(1).Infof(" info: the output version specified is invalid. %s\n", defaultVersionInfo)
```
Did you intend to have a leading space before info:?

ncdc replied:
It doesn't look necessary to me

juanvallejo replied:
@ncdc Okay, updated this here https://github.com/kubernetes/kubernetes/pull/34763/commits/6fd704c174df46cae0a6d4a4b13f2ed76d358752

ncdc commented:
@fabianofranz review please? 😄

fabianofranz commented:
@k8s-bot ok to test

juanvallejo commented:
@fabianofranz I think there might have been a flake with the `Jenkins GCI GKE smoke e2e` test

fabianofranz commented:
LGTM, @ncdc of @brendandburns mind tagging?

juanvallejo committed:
log info on invalid --output-version

Object versions default to the current version (v1) when a specified
`--output-version` is invalid. This patch logs a warning when this is
the case. Cases affected are all commands with the `--output-version`
option, and anywhere runtime objects are converted to versioned objects.

**Example**
```
$ kubectl get pod <mypod> -o json --output-version=invalid
W1013 17:24:16.810278   26719 result.go:238]  info: the output version
specified (invalid) is invalid, defaulting to v1
{
        "kind": "Pod",
            "apiVersion": "v1",
                "metadata": {
                            "name": "mypod",
                                    "namespace": "test",
...
```

juanvallejo commented:
@fabianofranz I believe  Jenkins GCE e2e and Jenkins GCE etcd3 e2e are flakes

ncdc commented:
@k8s-bot cvm gce e2e test this

ncdc commented:
@k8s-bot gce etcd3 e2e test this

ncdc commented on pkg/kubectl/resource/result.go:
```diff
@@ -229,6 +231,17 @@ func AsVersionedObject(infos []*Info, forceList bool, version schema.GroupVersio
 		object = converted
 	}
+
+	// validSpecifiedVersion resolves to true if the version passed to this function matches the
```
This comment looks more like godoc for a function than an inline comment. PTAL?

juanvallejo replied:
@ncdc Thanks for catching that. Fixed!

juanvallejo commented:
@ncdc or @brendandburns mind tagging once more? thanks!
